File: palin_sums.py
# problem 125
from math import sqrt

limit = 10000

def make_sum_consecutive_squares(limit):
	squares = [x*x for x in xrange(int(sqrt(limit)) + 1)]
	memo = {}
	def sum_consecutive_squares(n):
		for biggest in xrange(int(n**.5) + 1, 1, -1):
			for series in xrange(2, biggest):
				if (biggest-series, biggest) not in memo:
					memo[(biggest-series, biggest)] = sum(squares[biggest-series:biggest])
				series_sum = memo[(biggest-series, biggest)]
				if series_sum > n:
					break
				if series_sum == n:
					return True
		return False
	return sum_consecutive_squares

def is_palindrome(n):
	str_n = str(n)
	return str_n == str_n[::-1]

def check(n):
	return is_palindrome(n) and sum_consecutive_squares(n)

sum_consecutive_squares = make_sum_consecutive_squares(limit)
print sum(x for x in xrange(limit) if check(x))

File: palin_sums.cpp
// problem 125
// correct solution in 40m :(
#include <iostream>
#include <sstream>
#include <math.h>
#include <algorithm>
#include <string>
using namespace std;

const int limit = 100000000;
const int sqrt_limit = 10000;

static int squares[sqrt_limit + 1] = {0};

bool check(int n);
bool is_palindrome(int n);
bool sum_consecutive_squares(int n);

int main(void) {
	for (int i = 0; i < sqrt_limit + 1; ++i) {
		squares[i] = i*i;
	}
	long long int sum = 0;
	for (int i = 0; i < limit; ++i) {
		//cout << i << endl;
		if (check(i)) {
			sum += i;
		}
	}
	cout << sum << endl;
	return 0;
}

bool sum_consecutive_squares(int n) {
	for (int biggest = sqrt(n) + 1; biggest > 1; --biggest) {
		for (int series = 2; series < biggest; ++series) {
			int series_sum = 0;
			for (int i = biggest - series; i < biggest; ++i) {
				series_sum += squares[i];
			}
			if (series_sum > n) break;
			if (series_sum == n) return true;
		}
	}
	return false;
}

bool is_palindrome(int n) {
	int m = n, r = 0;
	while (m) {
		r = r*10 + m%10;
		m /= 10;
	}
	return n == r;
}

bool check(int n) {
	return is_palindrome(n) && sum_consecutive_squares(n);
}